Mathlib/NumberTheory/ModularForms/EisensteinSeries/UniformConvergence.lean

```diff
@@ -0,0 +1,291 @@
+of weight `k` and level `Γ(N)` with congruence condition `a : Fin 2 → ZMod N`.
+-/
+
+noncomputable section
+
+open Complex UpperHalfPlane Set Finset
+
+open scoped BigOperators Matrix UpperHalfPlane Complex
+
+namespace EisensteinSeries
+
+section bounding_functions
+
+/-- Auxiliary function used for bounding Eisenstein series. -/
+def r1 (z : ℍ) : ℝ := z.im ^ 2 / (z.re ^ 2 + z.im ^ 2)
+
+lemma r1' (z : ℍ) : r1 z = 1 / ((z.re / z.im) ^ 2 + 1) := by
+  field_simp [r1, im_pos z]
+
+theorem r1_pos (z : ℍ) : 0 < r1 z := by
+  dsimp only [r1]
+  positivity
+
+/-- This function is used to give an upper bound on Eisenstein series. -/
+def r (z : ℍ) : ℝ := min z.im (Real.sqrt (r1 z))
+
+lemma r_pos (z : ℍ) : 0 < r z := by
+  simp only [r, lt_min_iff, im_pos, Real.sqrt_pos, r1_pos, and_self]
+
+lemma r1_aux_bound (z : ℍ) (δ : ℝ) {ε : ℝ} (hε : 1 ≤ ε^2) :
+    (z.im ^ 2) / (z.re ^ 2 + z.im ^ 2) ≤ (δ * z.re + ε) ^ 2 + (δ * z.im) ^ 2 := by
+  have H1 : (δ * z.re + ε) ^ 2 + (δ * z.im) ^ 2 =
+        δ ^ 2 * (z.re ^ 2 + z.im ^ 2) + ε * 2 * δ * z.re + ε^2 := by ring
+  have H2 : (δ ^ 2 * (z.re ^ 2 + z.im ^ 2) + ε * 2 * δ * z.re + ε^2) * (z.re ^ 2 + z.im ^ 2)
+    - (z.im ^ 2) = (δ * (z.re ^ 2 + z.im ^ 2)+ ε * z.re)^2 + (ε^2 - 1)* (z.im)^2 := by ring
+  rw [H1, div_le_iff, ← sub_nonneg, H2]
+  · apply add_nonneg (pow_two_nonneg _) ?_
+    apply mul_nonneg
+    · linarith
+    · apply pow_two_nonneg
+  · positivity
+
+lemma auxbound1 (z : ℍ) {δ : ℝ} (ε : ℝ) (hδ : 1 ≤ δ ^ 2) : r z ≤ Complex.abs (δ * (z : ℂ) + ε) := by
+  rw [r, Complex.abs]
+  have H1 : (z : ℂ).im ≤
+    Real.sqrt ((δ * (z : ℂ).re + ε) * (δ * (z : ℂ).re + ε) + (δ * z : ℂ).im * (δ * z : ℂ).im) := by
+    have h1 : (δ * z : ℂ).im * (δ * z : ℂ).im = δ ^ 2 * (z : ℂ).im * (z : ℂ).im := by
+      simp only [mul_im, ofReal_re, coe_im, ofReal_im, coe_re, zero_mul, add_zero]
+      ring
+    rw [Real.le_sqrt', h1]
+    · nlinarith
+    · exact z.2
+  simp only [UpperHalfPlane.coe_im, UpperHalfPlane.coe_re, AbsoluteValue.coe_mk, MulHom.coe_mk,
+    min_le_iff] at *
+  left
+  simp only [one_le_sq_iff_one_le_abs, mul_im, ofReal_re, coe_im, ofReal_im, coe_re, zero_mul,
+    add_zero, normSq_apply, add_re, mul_re, sub_zero, add_im] at *
+  exact H1
+
+lemma auxbound2 (z : ℍ) (δ : ℝ) {ε : ℝ} (hε : 1 ≤ ε ^ 2) : r z ≤ Complex.abs (δ * (z : ℂ) + ε) := by
+  rw [r, Complex.abs, min_le_iff]
+  have H1 : Real.sqrt (r1 z) ≤ Real.sqrt ((δ * (z : ℂ).re + ε) * (δ * (z : ℂ).re + ε) +
+      δ * (z : ℂ).im * (δ * (z : ℂ).im)) := by
+    rw [r1, Real.sqrt_le_sqrt_iff, ← pow_two, ← pow_two]
+    apply r1_aux_bound z δ hε
+    nlinarith
+  right
+  simp only [ne_eq, coe_re, coe_im, normSq_apply, AbsoluteValue.coe_mk, MulHom.coe_mk, add_re,
+    mul_re, ofReal_re, ofReal_im, zero_mul, sub_zero, add_im, mul_im, add_zero] at *
+  exact H1
+
+lemma ne_zero_if_max {x : Fin 2 → ℤ} (hx : x ≠ 0)
+    (h : max (x 0).natAbs (x 1).natAbs = (x 0).natAbs) : x 0 ≠ 0 := by
+  contrapose! hx
+  ext x
+  fin_cases x <;> aesop
```

Comment:
One line longer, but twice as fast:
```suggestion
lemma left_ne_zero_of_max_eq {x : Fin 2 → ℤ} (hx : x ≠ 0)
    (h : max (x 0).natAbs (x 1).natAbs = (x 0).natAbs) : x 0 ≠ 0 := by
  contrapose! hx
  rw [hx, Int.natAbs_zero, max_eq_left_iff, Nat.le_zero, Int.natAbs_eq_zero] at h
  ext1 j
  fin_cases j <;> assumption
```
(If you prefer the shorter proof I won't insist; but please don't use `x` for the name of the variable introduced by `ext`, since `x` is already the name of the function, so you're trying to reason about the function `x` evaluated at `x`!)